include/libpmemobj++/experimental/mpsc_queue.hpp

```diff
@@ -156,27 +195,126 @@ class mpsc_queue {
 		bool
 		try_produce(size_t size, Function &&f)
 		{
-			entry dram_entry;
-			assert(size <= CACHELINE_SIZE);
+			auto data = std::unique_ptr<char[]>(new char[size]);
+			auto range = pmem::obj::slice<char *>(
+				data.get(), data.get() + size);
+
+			auto req_size = pmem::detail::align_up(
+				size + sizeof(first_block::size),
+				CACHELINE_SIZE);
 			auto offset = ringbuf_acquire(queue->ring_buffer.get(),
-						      w, CACHELINE_SIZE);
-			if (offset != -1) {
-				char *data = queue->buf + offset;
-				auto range = pmem::obj::slice<char *>(
-					dram_entry.data,
-					dram_entry.data + size);
-				f(range);
-				dram_entry.size = size;
-				pmemobj_memcpy(queue->pop.handle(), data,
-					       (char *)&dram_entry,
-					       CACHELINE_SIZE,
-					       PMEMOBJ_F_MEM_NONTEMPORAL);
-				pmemobj_drain(queue->pop.handle());
+						      w, req_size);
+
+			if (offset == -1)
+				return false;
+
+			f(range);
+
+			store_to_log(pmem::obj::string_view(data.get(), size),
+				     queue->buf + offset);
+
+			ringbuf_produce(queue->ring_buffer.get(), w);
+
+			return true;
+		}
+
+		bool
+		try_produce(pmem::obj::string_view data)
+		{
+			auto req_size = pmem::detail::align_up(
+				data.size() + sizeof(first_block::size),
+				CACHELINE_SIZE);
+			auto offset = ringbuf_acquire(queue->ring_buffer.get(),
+						      w, req_size);
+
+			if (offset == -1)
+				return false;
+
+			store_to_log(data, queue->buf + offset);
+
+			ringbuf_produce(queue->ring_buffer.get(), w);
 
-				ringbuf_produce(queue->ring_buffer.get(), w);
-				return true;
+			return true;
+		}
+
+	private:
+		void
+		store_to_log(pmem::obj::string_view data, char *log_data)
```

Comment:
This algorithm in libpmemobj also uses a checksum to verify consistency of the log. If you don't have that checksum and you don't care about failure atomicity then this can be simplified to just aligning down to a cache line, doing a copy, and then copying the remainder.